Add GraphQL types for document data

eugenk opened the pull request:
This adds all the models to GraphQL API. It depends on ~~https://github.com/ontohub/ontohub-models/pull/122~~ and ~~#279~~.

eugenk commented:
Hold on, there are still query specs missing. I will add them by the end of the week.

eugenk commented:
Now this is almost ready. It only needs to wait for #279. I will review that one tomorrow.

eugenk commented:
Rebased on master.

phyrog commented:
Is it ready for review now?

eugenk commented:
I just rebased on master. Now, this is ready for review.

phyrog commented:
I was able to reproduce the error consistently by rebasing this branch on master and running

```sh
rspec --seed 34649 spec/graphql/types/language_type_spec.rb
```

phyrog commented:
The problem is probably a difference in used locales. I'm not sure how to find out the exactly used values, but Ruby and Postgres both sort according to the used language (collation), they seem to differ on my system and on Travis. This results in a difference when comparing two lists, one sorted by Ruby, the other sorted by Postgres.

I'm using en_GB as my systems language, so one or the other probably uses this and the other uses en_US or something.

One workaround might be to use Postgres for both sortings (so you'd need to change e.g. `language_type_spec.rb:30`).

eugenk commented:
A lot of new errors appeared after rebasing on master. I will debug this tomorrow. Good to know that  the RSpec seed can be used to reproduce FactoryBot and Faker behaviour.

eugenk commented:
By the way, I could not reproduce the error on my machine with the seed 34649 after rebasing on master.

eugenk commented:
That's difficult to debug because I cannot reproduce it on my machine. Even setting `LC_ALL=en_US.UTF-8` (my default is `LC_ALL=en_GB.UTF-8`) doesn't change a thing in either application (ruby, postgres).

phyrog commented:
Did you get the same results for the commands above or do you get a consistent sorting?

eugenk commented:
I get consistent sorting (`foo-z` < `foom`). That's why I struggle debugging this.

phyrog commented:
How about `echo "foo-z\nfoom" | sort`? I'm getting `foom` < `foo-z` here.

eugenk commented:
Results in `foo-z` < `foom`

phyrog commented:
Can you run `locale` (I'm not sure if Mac has this)?

My output is:

```
LANG=en_GB.UTF-8
LC_CTYPE="en_GB.UTF-8"
LC_NUMERIC="en_GB.UTF-8"
LC_TIME="en_GB.UTF-8"
LC_COLLATE="en_GB.UTF-8"
LC_MONETARY="en_GB.UTF-8"
LC_MESSAGES="en_GB.UTF-8"
LC_PAPER="en_GB.UTF-8"
LC_NAME="en_GB.UTF-8"
LC_ADDRESS="en_GB.UTF-8"
LC_TELEPHONE="en_GB.UTF-8"
LC_MEASUREMENT="en_GB.UTF-8"
LC_IDENTIFICATION="en_GB.UTF-8"
LC_ALL=
```

eugenk commented:
```
 ❯ locale
LANG=
LC_COLLATE="en_GB.UTF-8"
LC_CTYPE="en_GB.UTF-8"
LC_MESSAGES="en_GB.UTF-8"
LC_MONETARY="en_GB.UTF-8"
LC_NUMERIC="en_GB.UTF-8"
LC_TIME="en_GB.UTF-8"
LC_ALL="en_GB.UTF-8
```

phyrog commented:
Can you check the output of the above commands on a Linux PC (or VM)? I'm kind of running out of ideas...

eugenk commented:
On my linux machine:
```
 ❯ ruby -e 'puts ["foo-z", "foom"].sort'
foo-z
foom

 ❯ echo "foo-z\nfoom" | sort
foom
foo-z

 ❯ locale
LANG=en_GB.UTF-8
LC_CTYPE="en_GB.UTF-8"
LC_NUMERIC="en_GB.UTF-8"
LC_TIME="en_GB.UTF-8"
LC_COLLATE="en_GB.UTF-8"
LC_MONETARY="en_GB.UTF-8"
LC_MESSAGES="en_GB.UTF-8"
LC_PAPER="en_GB.UTF-8"
LC_NAME="en_GB.UTF-8"
LC_ADDRESS="en_GB.UTF-8"
LC_TELEPHONE="en_GB.UTF-8"
LC_MEASUREMENT="en_GB.UTF-8"
LC_IDENTIFICATION="en_GB.UTF-8"
LC_ALL=en_GB.UTF-8
```
This is weird.

phyrog commented:
I just found [this link](https://simply.name/pg-lc-collate.html), so the problem seems to be known at least.

eugenk commented:
Just for documentation:
```text
 ❯ psql -U postgres -c "SELECT name, setting FROM pg_settings WHERE category ~ 'Locale';"
Timing is on.
Pager is always used.
            name            |      setting
----------------------------+--------------------
 client_encoding            | UTF8
 DateStyle                  | ISO, DMY
 default_text_search_config | pg_catalog.english
 extra_float_digits         | 0
 IntervalStyle              | postgres
 lc_collate                 | en_GB.UTF-8
 lc_ctype                   | en_GB.UTF-8
 lc_messages                | en_GB.UTF-8
 lc_monetary                | en_GB.UTF-8
 lc_numeric                 | en_GB.UTF-8
 lc_time                    | en_GB.UTF-8
 server_encoding            | UTF8
 TimeZone                   | Europe/Berlin
 timezone_abbreviations     | Default
(14 rows)
```

phyrog commented:
```
❯ psql -U postgres -c "SELECT name, setting FROM pg_settings WHERE category ~ 'Locale';"
            name            |      setting       
----------------------------+--------------------
 client_encoding            | UTF8
 DateStyle                  | ISO, MDY
 default_text_search_config | pg_catalog.english
 extra_float_digits         | 0
 IntervalStyle              | postgres
 lc_collate                 | en_US.UTF-8
 lc_ctype                   | en_US.UTF-8
 lc_messages                | en_US.UTF-8
 lc_monetary                | en_US.UTF-8
 lc_numeric                 | en_US.UTF-8
 lc_time                    | en_US.UTF-8
 server_encoding            | UTF8
 TimeZone                   | Europe/Berlin
 timezone_abbreviations     | Default
(14 rows)
```

`en_US` because it seems I used `en_US` when I first initialized the postgres database. However `en_US` and `en_GB` both sort `foom` and `foo-z` the same, so that's not the problem.

eugenk commented:
Your link (https://github.com/ontohub/ontohub-backend/pull/276#issuecomment-344378998) contained the solution: Use the locale "C" for sorting. I think Ruby uses glibc for `sort`, which sorts by looking at the bits only (without considering language-specifics). We can (and should) do the same in SQL - but only for the parameterised columns (`slug`, `loc_id`).

I added a few empty commits to trigger several travis builds in order to make sure that the green tests are not just a coincidence. I will delete them all.

phyrog commented:
Can reproduce the error with seed `20357`, collation seems to be set correctly:

```
psql -U postgres -d ontohub_development -c 'select table_schema, table_name, column_name, collation_name from information_schema.columns where collation_name is not null order by table_schema, table_name, ordinal_position;'

 table_schema |      table_name      | column_name | collation_name 
--------------+----------------------+-------------+----------------
 public       | languages            | slug        | C
 public       | loc_id_bases         | loc_id      | C
 public       | logic_mappings       | slug        | C
 public       | logics               | slug        | C
 public       | organizational_units | slug        | C
 public       | reasoners            | slug        | C
 public       | repositories         | slug        | C
 public       | serializations       | slug        | C
(8 rows)
```

eugenk commented:
There were a few bugs in there: I sort by name in the implementation and I sort by loc_id in the test. I think I have fixed them now (by rebase).

phyrog commented:
Btw I had to run `psql -U postgres -c "create collation C (locale = 'C');"` before this worked:

```
❯ psql -U postgres -c "select a from (values ('foo-z' collate C), ('foom' collate C)) as x(a) order by x.a;"                                       backend/git/add_graphql_types_for_document_data 
   a   
-------
 foo-z
 foom
(2 rows)

❯ psql -U postgres -c "select a from (values ('foo-z'), ('foom')) as x(a) order by x.a;"                                                           backend/git/add_graphql_types_for_document_data 
   a   
-------
 foom
 foo-z
(2 rows)
```

Test is still failing though...

eugenk commented:
I force-pushed a few minutes ago. Is the test still failing for you?

phyrog commented:
Not anymore as far as I can tell. What did you change?

eugenk commented:
There were several places like this:
https://github.com/ontohub/ontohub-backend/blob/fa0a9bc3ecd43bc2c473d2e5031ed8397e00df5e/app/graphql/types/premise_selection_type.rb#L29-L33
where it said `order(Sequel[:loc_id_bases][:name])` previously.

Also, I integrated https://github.com/ontohub/ontohub-models/pull/127

eugenk commented:
I removed the empty commits and updated the Gemfile-commit.

phyrog commented on app/graphql/types/query_type.rb:
```diff
@@ -57,4 +57,148 @@
+      description 'The id of the SignatureMorphism'
+    end
+
+    resolve(lambda do |_root, arguments, _context|
```
Needs authorization I think.

phyrog commented on app/graphql/types/query_type.rb:
```diff
@@ -57,4 +57,148 @@
+      description 'The id of the ReasonerConfiguration'
+    end
+
+    resolve(lambda do |_root, arguments, _context|
```
Needs authorization I think.

phyrog commented on app/graphql/types/query_type.rb:
```diff
@@ -57,4 +57,148 @@
+      description 'The id of the GeneratedAxiom'
+    end
+
+    resolve(lambda do |_root, arguments, _context|
```
Needs authorization I think.

phyrog commented on app/graphql/types/query_type.rb:
```diff
@@ -57,4 +57,148 @@
+      description 'The id of the PremiseSelection'
+    end
+
+    resolve(lambda do |_root, arguments, _context|
```
Needs authorization I think.

eugenk commented on app/graphql/types/query_type.rb:
```diff
@@ -118,6 +118,18 @@
+      description 'The id of the ReasoningAttempt'
+    end
+
+    resolve(lambda do |_root, arguments, _context|
```
It cannot be reached because it's only nested inside repository, but you are right. We should call `authorize` again to be sure. Maybe we change the nesting some day and forget to `authorize`.

eugenk replied:
Wait... this is not always nested in repository. It *needs* `authorize`.

eugenk replied:
This needs some work (additional associations) in the models...

eugenk commented on app/graphql/types/reasoning_status_enum.rb:
```diff
@@ -0,0 +1,19 @@
+  name 'ReasoningStatus'
+  description 'Specifies the reasoning status of a Conjecture'
+
+  value 'OPN', 'Open: No reasoning attempt has been finished.'
```
These abbreviations are standardised in the SZS ontology. I would not change them to full words. We can display the full words additionally in the frontend, but the abbreviations must be there.

phyrog replied:
Alright then

eugenk commented on app/graphql/types/symbol_type.rb:
```diff
@@ -0,0 +1,71 @@
+    end)
+  end
+
+  field :kind, !types.String do
```
There is no fixed set of kinds. Each language defines its own kinds and there are even languages that allow user-defined kinds. This cannot become an `enum`.

eugenk commented:
I *only* rebased on master. There was a minor conflict in the commit "Rename subject to commit".

eugenk commented:
I rebased on master and added authorisation with the "Add authorization to * query" commits. Now, this depends on https://github.com/ontohub/ontohub-models/pull/135.

eugenk commented:
The commit "Update ontohub-models." is now setting the current master. This is ready for the (final?) review now.

phyrog commented:
I think you missed the two comments in `document_type.rb` (they are shown as outdated because it was before some rebase).

eugenk commented on app/graphql/types/document_type.rb:
```diff
@@ -31,4 +31,13 @@
     end)
   end
+
+  field :importedBy, !types[!Types::DocumentType] do
```
Oh, this should be done. Thanks!

tillmo commented on app/graphql/types/mapping_origin_enum.rb:
```diff
@@ -0,0 +1,29 @@
+  name 'MappingOrigin'
+  description 'Specifies the origin of the Mapping'
+
+  # TODO: Till, please review and supply descriptions.
```
better description: "Specifies the origin (in the DOL abstract syntax) of the Mapping"

tillmo commented on app/graphql/types/mapping_type_enum.rb:
```diff
@@ -0,0 +1,29 @@
+  name 'MappingType'
+  description 'Specifies the type of the Mapping'
+
+  # TODO: Till, please review and supply descriptions.
```
better description: "Specifies the type of the Mapping (=link in the development graph)"

tillmo commented on app/graphql/types/oms_origin_enum.rb:
```diff
@@ -0,0 +1,38 @@
+  name 'OMSOrigin'
+  description 'Specifies the origin of the OMS'
+
+  # TODO: Till, please review and supply descriptions.
```
better description: "Specifies the origin (in the DOL abstract syntax) of the OMS"

phyrog replied:
I think the todos were more about descriptions for the enum values, see e.g. https://github.com/ontohub/ontohub-backend/blob/338aa5cfbdf8b39c6a4381603e0313f145fa7cc9/app/graphql/types/oms_origin_enum.rb#L8-L37
(the empty strings are supposed to be descriptions)

tillmo commented on app/graphql/types/oms_type.rb:
```diff
@@ -0,0 +1,170 @@
+    description 'The origin of this OMS'
+  end
+
+  # TODO: Till, please review and supply descriptions.
```
where is the corresponding Haskell datatype?

eugenk replied:
It is built at
https://github.com/spechub/Hets/blob/9c217c0bf6d4b39f9f5eb721e19a9a3125998d60/Persistence/DevGraph.hs#L437-L456

eugenk commented:
Done in 75eb83d6 and de008d3a

tillmo commented on app/graphql/types/oms_type.rb:
```diff
@@ -0,0 +1,170 @@
+    property :label_has_hiding
+  end
+
+  # TODO: Till, please review and supply descriptions.
```
OK, I have filed a commit

eugenk commented:
Looks like this is ready for review now.

eugenk commented on app/graphql/types/mapping_type_enum.rb:
```diff
@@ -0,0 +1,28 @@
+
+Types::MappingTypeEnum = GraphQL::EnumType.define do
+  name 'MappingType'
+  description 'Specifies the type of the Mapping (=link in the development graph)'
```
No. See 55d0cc3

eugenk commented on app/graphql/types/mapping_origin_enum.rb:
```diff
@@ -0,0 +1,28 @@
+  value 'dg_link_view', 'interpretation (view)'
+  value 'dg_link_align', 'alignment'
+  value 'dg_link_fit_view', 'fitting view on an instantiation of a parameterised OMS (CASL syntax)'
+  value 'dg_link_fit_view_imp', 'implict fitting view on an instantiation of a parameterised OMS (CASL syntax)'
```
No. See 55d0cc3

eugenk commented on app/graphql/types/mapping_origin_enum.rb:
```diff
@@ -0,0 +1,28 @@
+  value 'dg_link_inst_arg', 'argument of a parameterised OMS (CASL syntax)'
+  value 'dg_link_view', 'interpretation (view)'
+  value 'dg_link_align', 'alignment'
+  value 'dg_link_fit_view', 'fitting view on an instantiation of a parameterised OMS (CASL syntax)'
```
No. See 55d0cc3

eugenk commented on app/graphql/types/mapping_origin_enum.rb:
```diff
@@ -0,0 +1,28 @@
+  value 'dg_link_closed_lenv', 'closure OMS'
+  value 'dg_link_imports', 'CASL imports'
+  value 'dg_link_intersect', 'intersection'
+  value 'dg_link_morph', 'morphism created during an instantiation of a parameterised OMS (CASL syntax)'
```
No. See 55d0cc3

phyrog commented:
Oops, sorry, I introduced some conflicts by merging #313. Also needs a `rake graphql:write`
